pkg/controller/endpointslice/reconciler.go

```diff
@@ -0,0 +1,288 @@
+		slice.Name = sliceName
+		slicesToUpdate = append(slicesToUpdate, slice)
+	}
+
+	for _, endpointSlice := range slicesToCreate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Create(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error creating EndpointSlice for Service %s/%s: %v", r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	for _, endpointSlice := range slicesToUpdate {
+		addTriggerTimeAnnotation(endpointSlice, triggerTime)
+		_, err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Update(endpointSlice)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error updating %s EndpointSlice for Service %s/%s: %v", endpointSlice.Name, r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	for sliceNamesToDelete.Len() > 0 {
+		sliceName, _ := sliceNamesToDelete.PopAny()
+		err := r.client.DiscoveryV1alpha1().EndpointSlices(r.service.Namespace).Delete(sliceName, &metav1.DeleteOptions{})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Error deleting %s EndpointSlice for Service %s/%s: %v", sliceName, r.service.Namespace, r.service.Name, err))
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (r *Reconciler) reconcileByPortHash(
+	existingSlices []discovery.EndpointSlice,
+	desiredEndpoints []discovery.Endpoint,
+	endpointMeta EndpointMeta,
+) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice, sets.String) {
+	// build data structures for desired state
+	endpointsByHash := map[string]discovery.Endpoint{}
+	desiredEndpointHashes := sets.String{}
+	for _, desiredEndpoint := range desiredEndpoints {
+		eHash := endpointToHash(desiredEndpoint)
+		desiredEndpointHashes.Insert(eHash)
+		endpointsByHash[eHash] = desiredEndpoint
+	}
+
+	// build data structures for existing state
+	existingEndpointHashes := sets.String{}
+	endpointHashesBySliceName := map[string]sets.String{}
+	for _, existingSlice := range existingSlices {
+		endpointHashesBySliceName[existingSlice.Name] = sets.String{}
+		for _, endpoint := range existingSlice.Endpoints {
+			eHash := endpointToHash(endpoint)
+			endpointsByHash[eHash] = endpoint
+			existingEndpointHashes.Insert(eHash)
+			endpointHashesBySliceName[existingSlice.Name].Insert(eHash)
+		}
+	}
+
+	epHashesToAdd := desiredEndpointHashes.Difference(existingEndpointHashes)
+	epHashesToRemove := existingEndpointHashes.Difference(desiredEndpointHashes)
+
+	sliceNamesToUpdate := sets.String{}
+	sliceNamesToDelete := sets.String{}
+
+	for name := range endpointHashesBySliceName {
+		// remove any endpoints that are no longer desired
+		if epHashesToRemove.Len() > 0 {
+			origLength := endpointHashesBySliceName[name].Len()
+			endpointHashesBySliceName[name] = endpointHashesBySliceName[name].Difference(epHashesToRemove)
+
+			// if slice has changed, add as many desired endpoints as it can fit
+			if endpointHashesBySliceName[name].Len() != origLength {
```

Comment:
I think you can keep this step simple by just removes unwanted endpoints from slices. 

Add new endpoints in the next step. `findSliceCandidate` helper function can prioritize EndpointSlice candidates: 
1. If EndpointSlice is already being updated
2. The fullest EndpointSlice

Reply:
I did a variation of this. I couldn't think of a way to implement a function like `findSliceCandidate` in a way that would work for this use case as well. The logic there is tied to finding enough room to fit endpoints in a single slice. The amount of endpoints in that case will always be < maxSliceSize. Here we're trying to fill up all slices marked for update as efficiently as possible. What I thought made the most sense here was to sort the EndpointSlices already being update in order of most full, and fill them up with desired endpoints. I think that will result in the fullest slices getting fuller as we want.